File: lib/rag/chunk.ts
import type { Chunk } from './types'

// Tunable. CJK is char-dense so a 400-char chunk is roughly equivalent to
// 200-300 English tokens — fits BGE-M3's optimal range and leaves room
// for top-k stacking in a single LLM call.
const TARGET = 420
const OVERLAP = 60
const MAX = 600  // hard cap; anything bigger gets split mid-sentence

// Split a single CJK-friendly text into roughly TARGET-sized chunks at clean boundaries.
// Boundary preference (descending): \n\n  >  。！？  >  ，；  >  any char.
function splitToChunks(text: string): string[] {
  const out: string[] = []
  let cursor = 0
  const len = text.length

  while (cursor < len) {
    const remaining = len - cursor
    if (remaining <= MAX) {
      out.push(text.slice(cursor))
      break
    }

    // Look for the best boundary in [cursor + TARGET, cursor + MAX].
    const windowStart = cursor + TARGET
    const windowEnd = Math.min(cursor + MAX, len)
    const window = text.slice(windowStart, windowEnd)

    let cut = -1
    // Prefer paragraph break.
    const para = window.indexOf('\n\n')
    if (para >= 0) cut = para + 2
    // Then sentence-end punct.
    if (cut < 0) {
      const m = window.match(/[。！？!?]/)
      if (m && m.index !== undefined) cut = m.index + 1
    }
    // Then comma / semicolon.
    if (cut < 0) {
      const m = window.match(/[，；,;]/)
      if (m && m.index !== undefined) cut = m.index + 1
    }
    // Hard cut if no boundary found.
    if (cut < 0) cut = window.length

    const end = windowStart + cut
    out.push(text.slice(cursor, end))
    // Step forward, leaving OVERLAP chars to preserve context across chunks.
    cursor = Math.max(end - OVERLAP, cursor + 1)
  }

  return out.map(s => s.trim()).filter(s => s.length >= 40)
}

// Parse a markdown file and produce labeled chunks. We track the nearest preceding ## heading
// so the UI can show "from §市场画像 in sea-mobile-2025.md" instead of just a chunk id.
export function chunkMarkdown(source: string, raw: string): Chunk[] {
  // Split text at headings while remembering them.
  const lines = raw.split('\n')
  const sections: { heading?: string; text: string; offset: number }[] = []
  let currentHeading: string | undefined
  let currentBody: string[] = []
  let currentOffset = 0
  let runningOffset = 0
  let sectionStartOffset = 0

  const flush = () => {
    const body = currentBody.join('\n').trim()
    if (body.length > 0) {
      sections.push({ heading: currentHeading, text: body, offset: sectionStartOffset })
    }
    currentBody = []
  }

  for (const line of lines) {
    const lineLen = line.length + 1  // +1 for the \n we split on
    if (/^##\s+/.test(line)) {
      flush()
      currentHeading = line.replace(/^##\s+/, '').trim()
      sectionStartOffset = runningOffset + lineLen
    } else if (/^#\s+/.test(line)) {
      // top-level # heading: treat as doc title, skip from body but remember not as section header
      flush()
      currentHeading = undefined
      sectionStartOffset = runningOffset + lineLen
    } else {
      if (currentBody.length === 0) sectionStartOffset = runningOffset
      currentBody.push(line)
    }
    runningOffset += lineLen
  }
  flush()

  const chunks: Chunk[] = []
  let idx = 0
  for (const sec of sections) {
    const parts = splitToChunks(sec.text)
    let partOffset = sec.offset
    for (const part of parts) {
      // Find where this part starts within sec.text to compute absolute offset.
      const localStart = sec.text.indexOf(part)
      const absStart = localStart >= 0 ? sec.offset + localStart : partOffset
      const absEnd = absStart + part.length
      chunks.push({
        id: `${source}#${idx}`,
        source,
        heading: sec.heading,
        text: part,
        startChar: absStart,
        endChar: absEnd,
      })
      idx += 1
      partOffset = absEnd
    }
  }
  return chunks
}

File: server/stratsquad/rag/chunk.py
"""CJK-aware markdown chunker. Matches the TS lib/rag/chunk.ts behavior."""
from __future__ import annotations
import re
from dataclasses import dataclass


TARGET = 420
OVERLAP = 60
MAX = 600


@dataclass
class Chunk:
    id: str
    source: str
    heading: str | None
    text: str
    start_char: int
    end_char: int


def _split_to_chunks(text: str) -> list[str]:
    out: list[str] = []
    cursor = 0
    n = len(text)
    while cursor < n:
        remaining = n - cursor
        if remaining <= MAX:
            out.append(text[cursor:])
            break
        win_start = cursor + TARGET
        win_end = min(cursor + MAX, n)
        window = text[win_start:win_end]
        cut = -1
        i = window.find("\n\n")
        if i >= 0:
            cut = i + 2
        if cut < 0:
            m = re.search(r"[。！？!?]", window)
            if m:
                cut = m.end()
        if cut < 0:
            m = re.search(r"[，；,;]", window)
            if m:
                cut = m.end()
        if cut < 0:
            cut = len(window)
        end = win_start + cut
        out.append(text[cursor:end])
        cursor = max(end - OVERLAP, cursor + 1)
    return [s.strip() for s in out if len(s.strip()) >= 40]


def chunk_markdown(source: str, raw: str) -> list[Chunk]:
    lines = raw.split("\n")
    sections: list[dict] = []
    current_heading: str | None = None
    current_body: list[str] = []
    running_offset = 0
    section_start = 0

    def flush() -> None:
        nonlocal current_body
        body = "\n".join(current_body).strip()
        if body:
            sections.append({"heading": current_heading, "text": body, "offset": section_start})
        current_body = []

    for line in lines:
        line_len = len(line) + 1
        if re.match(r"^##\s+", line):
            flush()
            current_heading = re.sub(r"^##\s+", "", line).strip()
            section_start = running_offset + line_len
        elif re.match(r"^#\s+", line):
            flush()
            current_heading = None
            section_start = running_offset + line_len
        else:
            if not current_body:
                section_start = running_offset
            current_body.append(line)
        running_offset += line_len
    flush()

    chunks: list[Chunk] = []
    idx = 0
    for sec in sections:
        parts = _split_to_chunks(sec["text"])
        for part in parts:
            local = sec["text"].find(part)
            abs_start = sec["offset"] + local if local >= 0 else sec["offset"]
            chunks.append(Chunk(
                id=f"{source}#{idx}",
                source=source,
                heading=sec["heading"],
                text=part,
                start_char=abs_start,
                end_char=abs_start + len(part),
            ))
            idx += 1
    return chunks
